### ISSUE TITLE
HTTPAuthSessionWrapper doesn't accept anonymous requests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3924]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| esteve reported|
|-|-|
|Trac ID|trac#3924|
|Type|defect|
|Created|2009-07-20 17:23:48Z|
|Branch|https://github.com/twisted/twisted/tree/anonymous-guard-3924|

HTTPAuthSessionWrapper won't accept anonymous requests, when an instance of twisted.cred.checkers.AllowAnonymousAccess (or any other checker that supports credentials.IAnonymous) is registered in a Portal.

Attachments:

* [3924-anonymous-checkers-httpauthsessionwrapper.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e5/1e5eec9b2eb80159009b4e2d2e3e90d890ac6d4f/e3a2f51ec124e1ff6548651335f19de054958255.patch) (5809 bytes) - added by esteve on 2009-07-20 18:30:12Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3924 3924
type__defect defect
reporter__esteve esteve
priority__normal normal
milestone__ 
branch__branches_anonymous_guard_3924 branches/anonymous-guard-3924
branch_author__esteve esteve
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1248110628000000 1248110628000000
changetime__1248629047000000 1248629047000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|esteve commented|
|-|-|

I've created a branch in Launchpad (lp:~esteve/twisted/3924-support-anonymous-checkers-in-httpauthsessionwrapper) with a patch. I'm writing a test right now :-)

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|esteve removed owner|
|-|-|

My patch changes HTTPAuthSessionWrapper#getChildWithDefault to return a util.DeferredResource, (which delegates authentication to the portal) if the HTTP request doesn't have an "authorization" header, instead of returning an instance of UnauthorizedResource. This makes it possible to use HTTPAuthSessionWrapper with a checker that handles IAnonymous credentials.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson commented|
|-|-|

This looks sane enough to me on a quick look.

This highly non-obvious URL may help someone else review the code:

http://bazaar.launchpad.net/~esteve/twisted/3924-support-anonymous-checkers-in-httpauthsessionwrapper/revision/15264?remember=15262&compare_revid=15262

(you could also 'propose a merge' in Launchpad, that would generate a diff for people to review).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [27187]) Branching to 'anonymous-guard-3924'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [27188]) Apply 3924-anonymous-checkers-httpauthsessionwrapper.patch

refs [#3924](https://github.com/twisted/twisted/issues/3924)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Patch looks really good.  There are a couple minor issues that I'm going to fix, then I'll merge it:

  1. I like to try to avoid using \-style line continuation.  It's not against the Twisted coding standard, and there are a lot of old examples of it in the codebase, but I dislike it. :)
  1. The whitespace change at the top of wrapper.py is just noise.
  1. The wrapper.py module docstring is now slightly wrong, it talks about how unauthenticated requests result in a 401; they are now delegated to the realm.
  1. `test_anonymousAccess` should be separated from the preceding test method by two blank lines, not one.
  1. `assertEquals` is preferred over `assertEqual`, for new code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [27192]) Merge anonymous-guard-3924

Author: esteve
Reviewer: exarkun
Fixes: [#3924](https://github.com/twisted/twisted/issues/3924)

Add support for anonymous sessions to `twisted.web.guard` by having the
wrapper attempt to log in to the portal using anonymous credentials
when a request without an authorization header is received.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
